drivers/adc/adc_max11102_17.c

```diff
@@ -0,0 +1,475 @@
+{
+	const struct max11102_17_config *config = dev->config;
+	struct max11102_17_data *data = dev->data;
+	int result;
+	uint8_t buffer_rx[1];
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+	struct spi_dt_spec bus;
+
+	memcpy(&bus, &config->bus, sizeof(bus));
+	bus.config.operation |= SPI_HOLD_ON_CS;
+
+	result = spi_read_dt(&bus, &rx);
+	if (result != 0) {
+		LOG_ERR("read failed with error %i", result);
+		return result;
+	}
+
+	gpio_pin_set_dt(&config->gpio_chsel, data->current_channel_id);
+
+	result = spi_read_dt(&config->bus, &rx);
+	if (result != 0) {
+		LOG_ERR("read failed with error %i", result);
+		return result;
+	}
+
+	return 0;
+}
+
+static int max11102_17_channel_setup(const struct device *dev,
+				     const struct adc_channel_cfg *channel_cfg)
+{
+	const struct max11102_17_config *config = dev->config;
+
+	LOG_DBG("read from ADC channel %i", channel_cfg->channel_id);
+
+	if (channel_cfg->reference != ADC_REF_EXTERNAL0) {
+		LOG_ERR("invalid reference %i", channel_cfg->reference);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("invalid gain %i", channel_cfg->gain);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("invalid acquisition time %i", channel_cfg->acquisition_time);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential != 0) {
+		LOG_ERR("differential inputs are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id > config->channel_count) {
+		LOG_ERR("invalid channel selection %i", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int max11102_17_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t necessary = sizeof(int16_t);
+
+	if (sequence->options) {
+		necessary *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < necessary) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int max11102_17_validate_sequence(const struct device *dev,
+					 const struct adc_sequence *sequence)
+{
+	const struct max11102_17_config *config = dev->config;
+	struct max11102_17_data *data = dev->data;
+	size_t sequence_channel_count = 0;
+	const size_t channel_maximum = 8*sizeof(sequence->channels);
+
+	if (sequence->resolution != config->resolution) {
+		LOG_ERR("invalid resolution");
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < channel_maximum; ++i) {
+		if ((BIT(i) & sequence->channels) == 0) {
+			continue;
+		}
+
+		if (i > config->channel_count) {
+			LOG_ERR("invalid channel selection");
+			return -EINVAL;
+		}
+
+		sequence_channel_count++;
+		data->sequence_channel_id = i;
+	}
+
+	if (sequence_channel_count == 0) {
+		LOG_ERR("no channel selected");
+		return -EINVAL;
+	}
+
+	if (sequence_channel_count > 1) {
+		LOG_ERR("multiple channels selected");
+		return -EINVAL;
```

Comment:
Why? Would it be that hard to support sequences with multiple channels?

Reply:
I thought this feature was intended for ADCs which can actually read out multiple channels at once, which this IC definitely cannot. Of course I could simulate this via multiple reads in a row.